pyfr/solvers/euler/kernels/rsolvers/exact.mako

```diff
@@ -0,0 +1,218 @@
+# -*- coding: utf-8 -*-
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+<%include file='pyfr.solvers.euler.kernels.primitives'/>
+
+<% gmrtg = (c['gamma'] - 1)/(2*c['gamma']) %>
+<% gprtg = (c['gamma'] + 1)/(2*c['gamma']) %>
+<% tgrgm = (2*c['gamma'])/(c['gamma'] - 1) %>
+<% tgrgp = (2*c['gamma'])/(c['gamma'] + 1) %>
+<% trgm = 2/(c['gamma'] - 1) %>
+<% trgp = 2/(c['gamma'] + 1) %>
+<% gmrgp = (c['gamma'] - 1)/(c['gamma'] + 1) %>
+<% hgm = 0.5*(c['gamma'] - 1) %>
+<% rgm = 1/(c['gamma'] - 1) %>
+<% gamma = c['gamma'] %>
+
+// Initial guess for pressure
+<%pyfr:macro name='init_p' params='rl, vl, pl, cl, rr, vr, pr, cr, p0'>
+    fpdtype_t bpv = 0.5*(pl + pr) + 0.125*(vl[0] - vr[0])*(rl + rr)*(cl + cr);
+    bpv = max(0.,bpv);
+    fpdtype_t pmin = min(pl,pr);
+    fpdtype_t pmax = max(pl,pr);
+    fpdtype_t rpmax = pmax/pmin;
+
+    if ((rpmax <= 2.) && (pmin <= bpv) && (bpv <= pmax)){
+        p0 = bpv;
+    }
+    else if (bpv < pmin){
+        // Two-rarefaction Riemann solve
+        fpdtype_t pre = pow(pl/pr,${gmrtg});
+        fpdtype_t um  = (pre*vl[0]/cl + vr[0]/cr + ${trgm}*(pre - 1.0))/(pre/cl + 1./cr);
+
+        fpdtype_t ptl = 1. - ${hgm}*(um - vl[0])/cl;
+        fpdtype_t ptr = 1. + ${hgm}*(um - vr[0])/cr;
+
+        p0 = 0.5*(pl*pow(ptl,${tgrgm}) + pr*pow(ptr,${tgrgm}));
+    }
+    else{
+        // Two-shock Riemann solve
+        fpdtype_t gl = sqrt((${trgp}/rl)/(${gmrgp}*pl + bpv));
+        fpdtype_t gr = sqrt((${trgp}/rr)/(${gmrgp}*pr + bpv));
+        p0 = (gl*pl + gr*pr - (vr[0] - vl[0]))/(gl + gr);
+    }
+</%pyfr:macro>
+
+// Star Flux, assuming covolume = 0. See Toro 2009 Eq.(4.86-4.87)
+<%pyfr:macro name='star_flux' params='p, ps, rs, cs, f, fd'>
+    if (p <= ps){
+       fpdtype_t pr = p/ps;
+       f  = ${trgm}*cs*(pow(pr,${gmrtg}) - 1.);
+       fd = pow(pr,${-gprtg})/(rs*cs);
+    }
+    else{
+       fpdtype_t as = ${trgp}/rs;
+       fpdtype_t bs = ${gmrgp}*ps;
+       fpdtype_t sapb = sqrt(as/(p + bs));
+       f  = (p - ps)*sapb;
+       fd = (1. - 0.5*(p - ps)/(p + bs))*sapb;
+    }
+</%pyfr:macro>
+
+// Primitive to inviscid flux
+// w = [density, v_1,..., v_ndims, p]^T
+<%pyfr:macro name='primitive_1dflux' params='w, f'>
+    fpdtype_t invrho = 1.0/w[0];
+
+    // Compute the velocities
+    fpdtype_t rhov[${ndims}];
+% for i in range(ndims):
+    rhov[${i}] = w[0]*w[${i + 1}];
+% endfor
+
+    // Compute the Energy
+    fpdtype_t E = w[${nvars-1}]*${rgm} + 0.5*invrho*(${pyfr.dot('rhov[{i}]', i=ndims)});
+
+    // Density and energy fluxes
+    f[0] = rhov[0];
+    f[${nvars - 1}] = (E + w[${nvars-1}])*w[1];
+
+    // Momentum fluxes
+% for i in range(ndims):
+% if i == 0:
+    f[${i+1}]= rhov[0]*w[${i+1}] + w[${nvars-1}];
+% else:
+    f[${i+1}]= rhov[0]*w[${i+1}];
+% endif
+% endfor
+</%pyfr:macro>
+
+
+// Exact solve solution decision tree
+<% switch = 0.0 %>
```

Comment:
Is this always 0?

Reply:
Toro has this value so it can be changed, I think under some circumstances you may want to change it.

Reply:
Is it something we want to expose in a config file?